subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publisher/AbstractMavenPublisher.java

```diff
@@ -47,24 +47,27 @@ public void publish(MavenNormalizedPublication publication, MavenArtifactReposit
             LOGGER.info("Publishing to repository '{}' ({})", artifactRepository.getName(), artifactRepository.getUrl());
         }
 
-        MavenPublishAction deployTask = createDeployTask(publication.getPomFile(), publication.getMetadataFile(), mavenRepositoryLocator, artifactRepository);
+        File pomFile = publication.getPomArtifact().getFile();
+        MavenPublishAction deployTask = createDeployTask(pomFile, mavenRepositoryLocator, artifactRepository);
         addPomAndArtifacts(deployTask, publication);
         execute(deployTask);
     }
 
-    abstract protected MavenPublishAction createDeployTask(File pomFile, File metadataFile, LocalMavenRepositoryLocator mavenRepositoryLocator, MavenArtifactRepository artifactRepository);
+    abstract protected MavenPublishAction createDeployTask(File pomFile, LocalMavenRepositoryLocator mavenRepositoryLocator, MavenArtifactRepository artifactRepository);
 
     private void addPomAndArtifacts(MavenPublishAction publishAction, MavenNormalizedPublication publication) {
         MavenArtifact mainArtifact = publication.getMainArtifact();
         if (mainArtifact != null) {
             publishAction.setMainArtifact(mainArtifact.getFile());
         }
 
-        for (MavenArtifact mavenArtifact : publication.getArtifacts()) {
-            if (mavenArtifact == mainArtifact) {
+        MavenArtifact pomArtifact = publication.getPomArtifact();
+
+        for (MavenArtifact artifact : publication.getAllArtifacts()) {
+            if (artifact == mainArtifact || artifact == pomArtifact) {
                 continue;
             }
-            publishAction.addAdditionalArtifact(mavenArtifact.getFile(), GUtil.elvis(mavenArtifact.getExtension(), ""), GUtil.elvis(mavenArtifact.getClassifier(), ""));
+            publishAction.addAdditionalArtifact(artifact.getFile(), GUtil.elvis(artifact.getExtension(), ""), GUtil.elvis(artifact.getClassifier(), ""));
```

Comment:
⭕️ Strings.nullToEmpty would be more efficient. `GUtil.elvis` was only added when we converted code from Groovy to Java and we needed to keep the exact semantics. It should be avoided where possible.